contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,118 @@
+
+On top of extra processing time, exports were taking up large amounts of our team’s time. Reliability needed fixing before we worked on performance, and firefighting always focused on reliability. More importantly, our team was spending large amounts of time firefighting and "babysitting" exports. They had to be reactive, rather than proactive.
+
+Simply, we outgrew the export system we put in place, and needed a more reliable, transparent, and cost-effective system if we wanted to scale further.
+
+## Temporal to the rescue
+
+Early in 2023, one of our engineers, [James](/community/profiles/90), opened an [RFC recommending using Temporal at PostHog](https://github.com/PostHog/meta/pull/99). Temporal is a workflow engine abstracting away the details of failure modes, retry logic, and timeouts. This enables developers to build and deploy rock-solid business logic. It is used by big companies with large amounts of data flows – like Uber, Coinbase, Doordash, and Hashicorp.
+
+In Q2, we set a goal of "rock solid-batch processing" and [chose Temporal](https://github.com/PostHog/meta/pull/99) to build it because it helped solve issues like:
+
+- Retry, resume, cancel, and timeout logic per activity and workflow
+- Provides assurances about running exactly-once
+- Visibility into the status of workflows, including errors and logging.
+
+Instead of building all this functionality ourselves, Temporal takes care of it for us. After some debate, we felt building a workflow engine would not be a competitive advantage for us. Nothing about what we build on Temporal is novel enough to justify an investment into a competing runtime (as satisfying as it would be to write).
+
+## How we built our new export system with Temporal
+
+We didn’t completely rebuild our export system; many of the pieces still worked, we just needed to modify them to work with Temporal’s scheduling and execution capabilities. This meant building a batch export abstraction layer on top of Temporal. Specifically, the abstraction layer is four models:
+
+- `BatchExportDestination`: How users configure where to export data their data.
+- `BatchExportSchedule`: Maps to a [Temporal Schedule](https://keithtenzer.com/temporal/temporal_schedules/). It stores Schedule configurations to pass to Temporal.
+- `BatchExport`: The union of a `BatchExportSchedule` with a `BatchExportDestination`. Defines a destination for exported data, and a schedule to run said exports.
+- `BatchExportRun`: A `BatchExport` for each triggered workflow. Users aren’t meant to create these, only Temporal. They mostly serve to report on individual workflow executions.
+
+Users create a `BatchExport` that includes a schedule (which validates against the `BatchExportSchedule`) and a destination (stored as a `BatchExportDestination`). Data then gets read from PostHog and processed to write to the destination.
```

Comment:
I'm working to get a PR in that adds a sequence diagram to our docs explaining this process: https://github.com/PostHog/posthog.com/pull/6566. Maybe useful to link to.